BrowserKit/Tests/TabDataStoreTests/TabDataStoreTests.swift

```diff
@@ -58,37 +81,50 @@ final class TabDataStoreTests: XCTestCase {
                     DispatchQueue.main.async {
                         XCTAssertTrue(FileManager.default.fileExists(atPath: backupPath.path))
                     }
+                    do {
+                        let data = try Data(contentsOf: backupPath)
+                        let backupWindowData = try JSONDecoder().decode(WindowData.self, from: data)
+                        DispatchQueue.main.async {
+                            XCTAssertEqual(backupWindowData.id, windowData.id)
+                        }
+                    } catch {
+                        DispatchQueue.main.async {
+                            XCTFail("can't read the backup")
+                        }
+                    }
                 } else {
-                    XCTFail("can't create the backup path")
+                    DispatchQueue.main.async {
+                        XCTFail("can't create the backup path")
+                    }
                 }
             }
         }
     }
 
-    // MARK: Fetching Data
-    func testFetchAllWindowsData() async throws {
-        await tabDataStore.clearAllWindowsData()
+    func testFetchAllWindowsData() {
         let windowData1 = self.createMockWindow()
         let windowData2 = self.createMockWindow()
 
         // Save the WindowData objects
         Task {
             await tabDataStore.saveWindowData(window: windowData1)
-        }
-        Task {
-            await tabDataStore.saveWindowData(window: windowData2)
             Task {
-                // Fetch all WindowData objects
-                let fetchedWindowsData = await tabDataStore.fetchAllWindowsData()
-                // Verify the fetched data
-                XCTAssertEqual(fetchedWindowsData.count, 2)
-                XCTAssertTrue(fetchedWindowsData.contains(where: { $0.id == windowData1.id }))
-                XCTAssertTrue(fetchedWindowsData.contains(where: { $0.id == windowData2.id }))
+                await tabDataStore.saveWindowData(window: windowData2)
+                Task {
+                    // Fetch all WindowData objects
+                    let fetchedWindowsData = await tabDataStore.fetchAllWindowsData()
+                    // Verify the fetched data
+                    DispatchQueue.main.async {
+                        XCTAssertEqual(fetchedWindowsData.count, 2)
+                        XCTAssertTrue(fetchedWindowsData.contains(where: { $0.id == windowData1.id }))
+                        XCTAssertTrue(fetchedWindowsData.contains(where: { $0.id == windowData2.id }))
+                    }
+                }
```

Comment:
As we discussed on the call, you shouldn't mix GDC and async tasks. I also don't think these checks need to be on the main thread. I think the issues you were seeing will go away if you make the functions async and remove the Task blocks.